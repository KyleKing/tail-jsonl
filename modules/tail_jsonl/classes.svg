<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN"
 "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<!-- Generated by graphviz version 2.43.0 (0)
 -->
<!-- Title: classes Pages: 1 -->
<svg width="788pt" height="259pt"
 viewBox="0.00 0.00 788.00 259.00" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<g id="graph0" class="graph" transform="scale(1 1) rotate(0) translate(4 255)">
<title>classes</title>
<polygon fill="white" stroke="transparent" points="-4,4 -4,-255 784,-255 784,4 -4,4"/>
<!-- tail_jsonl.config.Config -->
<g id="node1" class="node">
<title>tail_jsonl.config.Config</title>
<polygon fill="none" stroke="black" points="29,-166.5 29,-250.5 145,-250.5 145,-166.5 29,-166.5"/>
<text text-anchor="start" x="64" y="-235.3" font-family="Times,serif" font-size="14.00">Config</text>
<polyline fill="none" stroke="black" points="29,-227.5 145,-227.5 "/>
<text text-anchor="start" x="37" y="-212.3" font-family="Times,serif" font-size="14.00">keys</text>
<text text-anchor="start" x="37" y="-197.3" font-family="Times,serif" font-size="14.00">styles : Styles</text>
<polyline fill="none" stroke="black" points="29,-189.5 145,-189.5 "/>
<text text-anchor="start" x="84.5" y="-174.3" font-family="Times,serif" font-size="14.00"> </text>
</g>
<!-- tail_jsonl.config.Keys -->
<g id="node2" class="node">
<title>tail_jsonl.config.Keys</title>
<polygon fill="none" stroke="black" points="0,-0.5 0,-114.5 174,-114.5 174,-0.5 0,-0.5"/>
<text text-anchor="start" x="70" y="-99.3" font-family="Times,serif" font-size="14.00">Keys</text>
<polyline fill="none" stroke="black" points="0,-91.5 174,-91.5 "/>
<text text-anchor="start" x="8" y="-76.3" font-family="Times,serif" font-size="14.00">level : List[str]</text>
<text text-anchor="start" x="8" y="-61.3" font-family="Times,serif" font-size="14.00">message : List[str]</text>
<text text-anchor="start" x="8" y="-46.3" font-family="Times,serif" font-size="14.00">on_own_line : List[str]</text>
<text text-anchor="start" x="8" y="-31.3" font-family="Times,serif" font-size="14.00">timestamp : List[str]</text>
<polyline fill="none" stroke="black" points="0,-23.5 174,-23.5 "/>
<text text-anchor="start" x="84.5" y="-8.3" font-family="Times,serif" font-size="14.00"> </text>
</g>
<!-- tail_jsonl.config.Keys&#45;&gt;tail_jsonl.config.Config -->
<g id="edge1" class="edge">
<title>tail_jsonl.config.Keys&#45;&gt;tail_jsonl.config.Config</title>
<path fill="none" stroke="black" d="M87,-114.83C87,-127.69 87,-141.31 87,-154.04"/>
<polygon fill="black" stroke="black" points="87,-154.42 91,-160.42 87,-166.42 83,-160.42 87,-154.42"/>
<text text-anchor="middle" x="103.5" y="-136.8" font-family="Times,serif" font-size="14.00" fill="green">keys</text>
</g>
<!-- tail_jsonl._private.core.Record -->
<g id="node3" class="node">
<title>tail_jsonl._private.core.Record</title>
<polygon fill="none" stroke="black" points="192.5,-0.5 192.5,-114.5 529.5,-114.5 529.5,-0.5 192.5,-0.5"/>
<text text-anchor="start" x="335.5" y="-99.3" font-family="Times,serif" font-size="14.00">Record</text>
<polyline fill="none" stroke="black" points="192.5,-91.5 529.5,-91.5 "/>
<text text-anchor="start" x="305.5" y="-76.3" font-family="Times,serif" font-size="14.00">data : Dict, dict</text>
<text text-anchor="start" x="305.5" y="-61.3" font-family="Times,serif" font-size="14.00">level : str</text>
<text text-anchor="start" x="305.5" y="-46.3" font-family="Times,serif" font-size="14.00">message : str</text>
<text text-anchor="start" x="305.5" y="-31.3" font-family="Times,serif" font-size="14.00">timestamp : str</text>
<polyline fill="none" stroke="black" points="192.5,-23.5 529.5,-23.5 "/>
<text text-anchor="start" x="200.5" y="-8.3" font-family="Times,serif" font-size="14.00">from_line(data: Dict, config: Config): &#39;Record&#39;</text>
</g>
<!-- tail_jsonl._runtime_type_check_setup._RuntimeTypeCheckingModes -->
<g id="node4" class="node">
<title>tail_jsonl._runtime_type_check_setup._RuntimeTypeCheckingModes</title>
<polygon fill="none" stroke="black" points="548,-23 548,-92 780,-92 780,-23 548,-23"/>
<text text-anchor="start" x="556" y="-76.8" font-family="Times,serif" font-size="14.00">_RuntimeTypeCheckingModes</text>
<polyline fill="none" stroke="black" points="548,-69 780,-69 "/>
<text text-anchor="start" x="644" y="-53.8" font-family="Times,serif" font-size="14.00">name</text>
<polyline fill="none" stroke="black" points="548,-46 780,-46 "/>
<text text-anchor="start" x="574.5" y="-30.8" font-family="Times,serif" font-size="14.00">from_environment(): Self</text>
</g>
</g>
</svg>
