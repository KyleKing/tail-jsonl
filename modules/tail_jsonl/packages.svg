<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN"
 "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<!-- Generated by graphviz version 2.43.0 (0)
 -->
<!-- Title: packages Pages: 1 -->
<svg width="484pt" height="116pt"
 viewBox="0.00 0.00 484.00 116.00" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<g id="graph0" class="graph" transform="scale(1 1) rotate(0) translate(4 112)">
<title>packages</title>
<polygon fill="white" stroke="transparent" points="-4,4 -4,-112 480,-112 480,4 -4,4"/>
<!-- tail_jsonl -->
<g id="node1" class="node">
<title>tail_jsonl</title>
<polygon fill="none" stroke="black" points="80,-36 0,-36 0,0 80,0 80,-36"/>
<text text-anchor="middle" x="40" y="-14.3" font-family="Times,serif" font-size="14.00">tail_jsonl</text>
</g>
<!-- tail_jsonl._private -->
<g id="node2" class="node">
<title>tail_jsonl._private</title>
<polygon fill="none" stroke="black" points="241.5,-36 98.5,-36 98.5,0 241.5,0 241.5,-36"/>
<text text-anchor="middle" x="170" y="-14.3" font-family="Times,serif" font-size="14.00">tail_jsonl._private</text>
</g>
<!-- tail_jsonl._private.core -->
<g id="node3" class="node">
<title>tail_jsonl._private.core</title>
<polygon fill="none" stroke="black" points="330,-108 152,-108 152,-72 330,-72 330,-108"/>
<text text-anchor="middle" x="241" y="-86.3" font-family="Times,serif" font-size="14.00">tail_jsonl._private.core</text>
</g>
<!-- tail_jsonl.config -->
<g id="node4" class="node">
<title>tail_jsonl.config</title>
<polygon fill="none" stroke="black" points="476,-108 348,-108 348,-72 476,-72 476,-108"/>
<text text-anchor="middle" x="412" y="-86.3" font-family="Times,serif" font-size="14.00">tail_jsonl.config</text>
</g>
<!-- tail_jsonl.scripts -->
<g id="node5" class="node">
<title>tail_jsonl.scripts</title>
<polygon fill="none" stroke="black" points="394,-36 260,-36 260,0 394,0 394,-36"/>
<text text-anchor="middle" x="327" y="-14.3" font-family="Times,serif" font-size="14.00">tail_jsonl.scripts</text>
</g>
<!-- tail_jsonl.scripts&#45;&gt;tail_jsonl._private.core -->
<g id="edge1" class="edge">
<title>tail_jsonl.scripts&#45;&gt;tail_jsonl._private.core</title>
<path fill="none" stroke="black" d="M305.74,-36.3C294.91,-45.12 281.58,-55.97 269.84,-65.53"/>
<polygon fill="black" stroke="black" points="262.01,-71.9 266.93,-62.09 265.89,-68.74 269.77,-65.58 269.77,-65.58 269.77,-65.58 265.89,-68.74 272.61,-69.07 262.01,-71.9 262.01,-71.9"/>
</g>
<!-- tail_jsonl.scripts&#45;&gt;tail_jsonl.config -->
<g id="edge2" class="edge">
<title>tail_jsonl.scripts&#45;&gt;tail_jsonl.config</title>
<path fill="none" stroke="black" d="M348.01,-36.3C358.72,-45.12 371.89,-55.97 383.5,-65.53"/>
<polygon fill="black" stroke="black" points="391.23,-71.9 380.65,-69.01 387.37,-68.72 383.51,-65.54 383.51,-65.54 383.51,-65.54 387.37,-68.72 386.37,-62.06 391.23,-71.9 391.23,-71.9"/>
</g>
</g>
</svg>
