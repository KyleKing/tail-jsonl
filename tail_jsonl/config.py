"""Configuration."""

from functools import cached_property

from beartype import beartype
from beartype.typing import Dict, List
from pydantic import BaseModel, Field


class Styles(BaseModel):
    """Styles configuration.

    Refer to https://rich.readthedocs.io/en/latest/style.html for available style

    Inspired by: https://github.com/Delgan/loguru/blob/07f94f3c8373733119f85aa8b9ca05ace3325a4b/loguru/_defaults.py#L31-L73

    And: https://github.com/hynek/structlog/blob/bcfc7f9e60640c150bffbdaeed6328e582f93d1e/src/structlog/dev.py#L126-L141

    """

    timestamp: str = 'dim grey'
    message: str = ''

    level_error: str = 'bold red'
    level_warn: str = 'bold yellow'
    level_info: str = 'bold green'
    level_debug: str = 'bold blue'
    level_fallback: str = ''

    key: str = 'green'
    value: str = ''

    @cached_property
    def _level_lookup(self) -> Dict[str, str]:
        return {
            'ERROR': self.level_error,
            'WARNING': self.level_warn,
            'WARN': self.level_warn,
            'INFO': self.level_info,
            'DEBUG': self.level_debug,
        }

    @beartype
    def get_level_style(self, level: str) -> str:
        """Return the right style for the specified level."""
        return self._level_lookup.get(level.upper(), self.level_fallback)


class Keys(BaseModel):
    """Special Keys."""

    timestamp: List[str] = Field(default_factory=lambda: ['timestamp', 'record.time.repr'])
    level: List[str] = Field(default_factory=lambda: ['level', 'record.level.name'])
    message: List[str] = Field(default_factory=lambda: ['event', 'record.message'])

    on_own_line: List[str] = Field(default_factory=lambda: ['text', 'exception'])


class Config(BaseModel):
    """`tail-jsonl` config."""

    styles: Styles = Field(default_factory=Styles)
    keys: Keys = Field(default_factory=Keys)
