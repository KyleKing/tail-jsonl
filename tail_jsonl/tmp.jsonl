{"jsonrpc": "2.0", "method": "sum", "params": [ null, 1, 2, 4, false, true], "id": "1"}
{"jsonrpc": "2.0", "method": "notify_hello", "params": [7]}
{"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": "2"}
{"event": "Testing", "user_id": "UUID('f86f89aa-a260-431a-a3a6-7b37a695e012'), UUID('f86f89aa-a260-431a-a3a6-7b37a695e012'), UUID('f86f89aa-a260-431a-a3a6-7b37a695e012'), UUID('f86f89aa-a260-431a-a3a6-7b37a695e012'), UUID('f86f89aa-a260-431a-a3a6-7b37a695e012'), UUID('f86f89aa-a260-431a-a3a6-7b37a695e012')", "something": 123, "level": "debug", "timestamp": "2022-12-22T16:52:28.380253Z", "func_name": "<module>", "filename": "test.py", "lineno": 15, "exception": "line 1\n\t\tline 2 ....    \n\t...?"}
