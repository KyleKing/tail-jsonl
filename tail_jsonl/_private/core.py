"""Core print logic."""

import json
import logging
from copy import copy

import dotted
from beartype import beartype
from beartype.typing import Any, Dict, List, Optional
from corallium.loggers.rich_printer import rich_printer
from corallium.loggers.styles import get_level
from pydantic import BaseModel
from rich.console import Console

from ..config import Config


@beartype
def _dot_pop(data: Dict, key: str) -> Optional[str]:  # type: ignore[type-arg]
    value = dotted.get(data, key)
    if isinstance(value, str):
        dotted.remove(data, key)
        return value or None
    return None


@beartype
def _pop_key(data: Dict, keys: List[str], fallback: str) -> Any:  # type: ignore[type-arg]
    """Recursively pop each key while searching for a match."""
    try:
        key = keys.pop(0)
    except IndexError:
        return fallback
    else:
        return _dot_pop(data, key) or _pop_key(data, keys, fallback)


@beartype
def pop_key(data: Dict, keys: List[str], fallback: str) -> Any:  # type: ignore[type-arg]
    """Safely find the first key in the data or default to the fallback."""
    return _pop_key(data, copy(keys), fallback)


class Record(BaseModel):
    """Record Model."""

    timestamp: str
    level: str
    message: str
    data: Dict  # type: ignore[type-arg]

    @classmethod
    @beartype
    def from_line(cls, data: Dict, config: Config) -> 'Record':  # type: ignore[type-arg]
        """Extract Record from jsonl."""
        return cls(
            timestamp=pop_key(data, config.keys.timestamp, '<no timestamp>'),
            level=pop_key(data, config.keys.level, ''),
            message=pop_key(data, config.keys.message, '<no message>'),
            data=data,
        )


@beartype
def print_record(line: str, console: Console, config: Config) -> None:
    """Format and print the record."""
    try:
        record = Record.from_line(json.loads(line), config=config)
    except Exception:  # noqa: PIE786
        console.print(line.rstrip(), markup=False, highlight=False)  # Print the unmodified line
        return

    if (_this_level := get_level(name=record.level)) == logging.NOTSET and record.level:
        record.data['_level_name'] = record.level

    printer_kwargs = {
        'message': record.message,
        'is_header': False,
        '_this_level': _this_level,
        '_is_text': False,
        '_console': console,
        '_styles': config.styles,
        '_keys_on_own_line': config.keys.on_own_line,
        'timestamp': record.timestamp,
    }
    keys = set(printer_kwargs)
    rich_printer(
        **printer_kwargs,
        # Ensure that there is no repeat keyword arguments
        **{f'_{key}' if key in keys else key: value for key, value in record.data.items()},
    )
