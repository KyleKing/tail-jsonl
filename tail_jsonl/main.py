"""Public CLI interface."""

import fileinput

from beartype import beartype
from rich.console import Console

from ._private.config import Config
from ._private.core import print_record


@beartype
def main() -> None:
    """CLI Entrypoint."""
    console = Console()
    # PLANNED: Support configuration with argparse or environment variable
    config = Config()
    for line in fileinput.input():
        print_record(line, console, config)
