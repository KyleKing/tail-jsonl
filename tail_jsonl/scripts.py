"""Start the command line program."""

import argparse
import fileinput
import sys
from pathlib import Path

from beartype import beartype
from beartype.typing import Dict, Optional
from corallium.tomllib import tomllib
from rich.console import Console

from . import __version__
from ._private.core import print_record
from .config import Config


@beartype
def _load_config(config_path: Optional[str]) -> Config:
    """Load specified configuration file."""
    user_config: Dict = {}  # type: ignore[type-arg]
    if config_path:
        user_config = tomllib.loads(Path(config_path).read_text(encoding='utf-8'))
    return Config(**user_config)


@beartype
def start() -> None:  # pragma: no cover
    """CLI Entrypoint."""
    # PLANNED: Add a flag (--debug & store_true) to print debugging information

    parser = argparse.ArgumentParser(description='Pipe JSONL Logs for pretty printing')
    parser.add_argument(
        '-v', '--version', action='version',
        version=f'%(prog)s {__version__}', help="Show program's version number and exit.",
    )
    parser.add_argument('--config-path', help='Path to a configuration file')
    options = parser.parse_args(sys.argv[1:])
    sys.argv = sys.argv[:1]  # Remove CLI before calling fileinput

    config = _load_config(options.config_path)
    console = Console()
    for line in fileinput.input():
        print_record(line, console, config)
