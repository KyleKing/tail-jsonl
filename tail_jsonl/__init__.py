"""tail_jsonl."""

from warnings import filterwarnings

from beartype.roar import BeartypeDecorHintPep585DeprecationWarning
from loguru import logger  # noqa: F401

__version__ = '1.1.2'
__pkg_name__ = 'tail_jsonl'

# ====== Above is the recommended code from calcipy_template and may be updated on new releases ======

# FYI: https://github.com/beartype/beartype#are-we-on-the-worst-timeline
filterwarnings('ignore', category=BeartypeDecorHintPep585DeprecationWarning)

from .main import main  # noqa: F401
