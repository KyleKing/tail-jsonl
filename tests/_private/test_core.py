"""Test core logic.

Generated sample JSONL data with:

```py
from loguru import logger
from calcipy.log_helpers import activate_debug_logging

from tail_jsonl import __pkg_name__

activate_debug_logging(pkg_names=[__pkg_name__])

logger.debug("debug-level log", data={"key1": 123})
logger.info("info-level log")
logger.warning("warning-level log")
logger.error("error-level log")
try:
    1 // 0
except Exception:
    logger.exception("exception-level log")
```

"""

import json
import platform

import pytest
from beartype import beartype
from beartype.typing import List
from rich.console import Console

from tail_jsonl._private.core import print_record
from tail_jsonl.config import Config
from tests.configuration import TEST_DATA_DIR


@beartype
def read_logs() -> List[str]:
    return (TEST_DATA_DIR / 'logs.jsonl').read_text(encoding='utf-8').strip().split('\n')


LOGS = read_logs()


@pytest.mark.parametrize('logs_index', [*range(len(LOGS))])
def test_core(logs_index, snapshot, console: Console):
    """Smoketest core."""
    print_record(LOGS[logs_index], console, Config())

    result = console.end_capture()

    assert result.strip()
    assert '<no ' not in result
    if platform.system() != 'Windows':
        assert result == snapshot


def test_core_no_key_matches(console: Console):
    print_record('{"key": null}', console, Config())

    result = console.end_capture()

    assert result.strip() == '<no timestamp>               [NOTSET ] <no message> key=None'


def test_core_bad_json(console: Console):
    print_record('{"bad json": None}', console, Config())

    result = console.end_capture()

    assert result.strip() == '{"bad json": None}'


def test_core_wrap(console: Console):
    print_record(json.dumps(dict.fromkeys(range(3), '-' * 3)), console, Config())

    result = console.end_capture()

    if platform.system() != 'Windows':
        expected = '<no timestamp>               [NOTSET ] <no message> 0=--- 1=--- 2=---'
        assert result.strip() == expected
