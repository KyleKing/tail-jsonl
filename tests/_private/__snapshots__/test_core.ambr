# serializer version: 1
# name: test_core[0]
  '''
  2023-01-31 06:16:00.910671-05:00 [DEBUG  ] debug-level log record={'elapsed': {'repr': '0:00:00.340931', 'seconds': 
  0.340931}, 'exception': None, 'extra': {'data': {'key1': 123}}, 'file': {'name': 'dodo.py', 'path': 
  '/Users/kyleking/Developer/packages/tail-jsonl/dodo.py'}, 'function': '<module>', 'level': {'icon': '🐞', 'no': 10}, 
  'line': 31, 'module': 'dodo', 'name': 'dodo', 'process': {'id': 16268, 'name': 'MainProcess'}, 'thread': {'id': 
  8821244160, 'name': 'MainThread'}, 'time': {'timestamp': 1675160160.910671}}
   ∟ text: 2023-01-31 06:16:00.910 | DEBUG    | dodo:<module>:31 - debug-level log
  
  
  '''
# ---
# name: test_core[1]
  '''
  2023-01-31 06:16:00.910915-05:00 [INFO   ] info-level log record={'elapsed': {'repr': '0:00:00.341175', 'seconds': 
  0.341175}, 'exception': None, 'extra': {}, 'file': {'name': 'dodo.py', 'path': 
  '/Users/kyleking/Developer/packages/tail-jsonl/dodo.py'}, 'function': '<module>', 'level': {'icon': 'ℹ️', 'no': 20}, 
  'line': 32, 'module': 'dodo', 'name': 'dodo', 'process': {'id': 16268, 'name': 'MainProcess'}, 'thread': {'id': 
  8821244160, 'name': 'MainThread'}, 'time': {'timestamp': 1675160160.910915}}
   ∟ text: 2023-01-31 06:16:00.910 | INFO     | dodo:<module>:32 - info-level log
  
  
  '''
# ---
# name: test_core[2]
  '''
  2023-01-31 06:16:00.911034-05:00 [WARNING] warning-level log record={'elapsed': {'repr': '0:00:00.341294', 'seconds': 
  0.341294}, 'exception': None, 'extra': {}, 'file': {'name': 'dodo.py', 'path': 
  '/Users/kyleking/Developer/packages/tail-jsonl/dodo.py'}, 'function': '<module>', 'level': {'icon': '⚠️', 'no': 30}, 
  'line': 33, 'module': 'dodo', 'name': 'dodo', 'process': {'id': 16268, 'name': 'MainProcess'}, 'thread': {'id': 
  8821244160, 'name': 'MainThread'}, 'time': {'timestamp': 1675160160.911034}}
   ∟ text: 2023-01-31 06:16:00.911 | WARNING  | dodo:<module>:33 - warning-level log
  
  
  '''
# ---
# name: test_core[3]
  '''
  2023-01-31 06:16:00.911331-05:00 [ERROR  ] error-level log record={'elapsed': {'repr': '0:00:00.341591', 'seconds': 
  0.341591}, 'exception': None, 'extra': {}, 'file': {'name': 'dodo.py', 'path': 
  '/Users/kyleking/Developer/packages/tail-jsonl/dodo.py'}, 'function': '<module>', 'level': {'icon': '❌', 'no': 40}, 
  'line': 34, 'module': 'dodo', 'name': 'dodo', 'process': {'id': 16268, 'name': 'MainProcess'}, 'thread': {'id': 
  8821244160, 'name': 'MainThread'}, 'time': {'timestamp': 1675160160.911331}}
   ∟ text: 2023-01-31 06:16:00.911 | ERROR    | dodo:<module>:34 - error-level log
  
  
  '''
# ---
# name: test_core[4]
  '''
  2023-01-31 06:16:00.911499-05:00 [ERROR  ] exception-level log record={'elapsed': {'repr': '0:00:00.341759', 'seconds': 
  0.341759}, 'exception': {'type': 'ZeroDivisionError', 'value': 'integer division or modulo by zero', 'traceback': True},
  'extra': {}, 'file': {'name': 'dodo.py', 'path': '/Users/kyleking/Developer/packages/tail-jsonl/dodo.py'}, 'function': 
  '<module>', 'level': {'icon': '❌', 'no': 40}, 'line': 38, 'module': 'dodo', 'name': 'dodo', 'process': {'id': 16268, 
  'name': 'MainProcess'}, 'thread': {'id': 8821244160, 'name': 'MainThread'}, 'time': {'timestamp': 1675160160.911499}}
   ∟ text: 2023-01-31 06:16:00.911 | ERROR    | dodo:<module>:38 - exception-level log
  Traceback (most recent call last):
  
    File "/Users/kyleking/Developer/packages/tail-jsonl/.venv/bin/doit", line 8, in <module>
      sys.exit(main())
      │   │    └ <function main at 0x103c3a2a0>
      │   └ <built-in function exit>
      └ <module 'sys' (built-in)>
    File "/Users/kyleking/Developer/packages/tail-jsonl/.venv/lib/python3.11/site-packages/doit/__main__.py", line 8, in 
  main
      sys.exit(DoitMain().run(sys.argv[1:]))
      │   │    │              │   └ ['/Users/kyleking/Developer/packages/tail-jsonl/.venv/bin/doit', 'run', 'test']
      │   │    │              └ <module 'sys' (built-in)>
      │   │    └ <class 'doit.doit_cmd.DoitMain'>
      │   └ <built-in function exit>
      └ <module 'sys' (built-in)>
    File "/Users/kyleking/Developer/packages/tail-jsonl/.venv/lib/python3.11/site-packages/doit/doit_cmd.py", line 294, in
  run
      return command.parse_execute(args)
             │       │             └ ['test']
             │       └ <function Command.parse_execute at 0x1037e0fe0>
             └ <doit.cmd_run.Run object at 0x103876f50>
    File "/Users/kyleking/Developer/packages/tail-jsonl/.venv/lib/python3.11/site-packages/doit/cmd_base.py", line 150, in
  parse_execute
      return self.execute(params, args)
             │    │       │       └ ['test']
             │    │       └ {'dep_file': '.doit.db', 'backend': 'dbm', 'codec_cls': 'json', 'check_file_uptodate': 'md5', 
  'dodoFile': 'dodo.py', 'cwdPath...
             │    └ <function DoitCmdBase.execute at 0x1037e1da0>
             └ <doit.cmd_run.Run object at 0x103876f50>
    File "/Users/kyleking/Developer/packages/tail-jsonl/.venv/lib/python3.11/site-packages/doit/cmd_base.py", line 524, in
  execute
      self.loader.setup(params)
      │    │      │     └ {'dep_file': '.doit.db', 'backend': 'dbm', 'codec_cls': 'json', 'check_file_uptodate': 'md5', 
  'dodoFile': 'dodo.py', 'cwdPath...
      │    │      └ <function DodoTaskLoader.setup at 0x1037e1800>
      │    └ <doit.cmd_base.DodoTaskLoader object at 0x102f748d0>
      └ <doit.cmd_run.Run object at 0x103876f50>
    File "/Users/kyleking/Developer/packages/tail-jsonl/.venv/lib/python3.11/site-packages/doit/cmd_base.py", line 394, in
  setup
      self.namespace = dict(inspect.getmembers(loader.get_module(
      │    │                │       │          │      └ <function get_module at 0x102f6fce0>
      │    │                │       │          └ <module 'doit.loader' from 
  '/Users/kyleking/Developer/packages/tail-jsonl/.venv/lib/python3.11/site-packages/doit/loader.py'>
      │    │                │       └ <function getmembers at 0x102f1dbc0>
      │    │                └ <module 'inspect' from 
  '/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/insp...
      │    └ None
      └ <doit.cmd_base.DodoTaskLoader object at 0x102f748d0>
    File "/Users/kyleking/Developer/packages/tail-jsonl/.venv/lib/python3.11/site-packages/doit/loader.py", line 96, in 
  get_module
      return importlib.import_module(os.path.splitext(file_name)[0])
             │         │             │  │    │        └ 'dodo.py'
             │         │             │  │    └ <function splitext at 0x102d0cea0>
             │         │             │  └ <module 'posixpath' (frozen)>
             │         │             └ <module 'os' (frozen)>
             │         └ <function import_module at 0x102f1cc20>
             └ <module 'importlib' from 
  '/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/im...
  
    File 
  "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/importlib/__init__.py"
  , line 126, in import_module
      return _bootstrap._gcd_import(name[level:], package, level)
             │          │           │    │        │        └ 0
             │          │           │    │        └ None
             │          │           │    └ 0
             │          │           └ 'dodo'
             │          └ <function _gcd_import at 0x102c53d80>
             └ <module '_frozen_importlib' (frozen)>
  
    File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
    File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
    File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
    File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
    File "<frozen importlib._bootstrap_external>", line 940, in exec_module
    File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  
  > File "/Users/kyleking/Developer/packages/tail-jsonl/dodo.py", line 36, in <module>
      1 // 0
  
  ZeroDivisionError: integer division or modulo by zero
  
  
  '''
# ---
